addons/l10n_id/models/qris_transaction.py

```diff
@@ -0,0 +1,118 @@
+    #     export_string_translation=False,
+    # )
+    paid = fields.Boolean(help="Payment Status of QRIS")
+
+    @api.constrains('model')
+    def _constraint_model(self):
+        # only allow the models as defined in the MODEL_ID_FIELDS key
+        available_models = MODEL_ID_FIELDS.keys()
+        if self.model and self.model not in available_models:
+            raise ValidationError(_("QRIS capability is not extended to model %s yet!", self.model))
+
+    def _get_record(self):
+        """ Get the backend record that the qris transaction is handling """
+        self.ensure_one()
+        id_field = MODEL_ID_FIELDS.get(self.model)
+        if id_field:
+            model_id = int(self.model_id) if self.model == 'account.move' else self.model_id
+            return self.env[self.model].search([(id_field, '=', model_id)])
+        else:
+            return self.env[self.model]
+
+    @api.model
+    def _get_latest_transaction(self, model, model_id):
+        """ Find latest transaction associated to the model and model_id """
+        return self.search([('model', '=', model), ('model_id', '=', model_id)], order='qris_creation_datetime desc', limit=1)
+
+    def _l10n_id_get_qris_qr_statuses(self, invoice_bank_id):
+        """ Fetch the result of the transaction
+
+        :param invoice_bank_id (Model <res.partner.bank>): bank (with QRIS configuration)
+        :returns tuple(bool, dict): paid/unpaid status and status_response from QRIS
+        """
+        # storing all failure transactions in case final result is unpaid
+        unpaid_status_data = []
+
+        # Looping to make requests is far from ideal, but we have no choices as they don't allow getting multiple QR result at once.
+        # Ensure to loop in reverse and check from the most recent QR code.
+        for transaction in self.sorted(lambda t: t.qris_creation_datetime):
+            status_response = invoice_bank_id._l10n_id_qris_fetch_status(transaction)
+            if status_response['data'].get('qris_status') == 'paid':
+                transaction.paid = True
+                return {
+                    'paid': True,
+                    'qr_statuses': [status_response['data']]
+                }
+            else:
+                unpaid_status_data.append(status_response['data'])
+
+        return {
+            'paid': False,
+            'qr_statuses': unpaid_status_data
+        }
+
+    @api.autovacuum
+    def _gc_remove_pointless_qris_transactions(self):
+        """ Removing transactions that are no longer useful. These transactions include:
+
+        - All unpaid transactions whose latest qris_invoice_details is created > 30 minutes alrady
+        - Paid transactions (and optionally the original record's payment state is paid)
+            - for account.move it's payment_state == 'in'
+            - for pos.order, it's state in ('paid', 'done', 'invoiced')
+        """
+        transactions = self.env['l10n_id.qris.transaction'].search([])
```

Comment:
Please add the domain in the search right away, to avoid loading all of them at once.
unpaid transactions with a creation datetime <= now - x seconds